File: contest-management-java/src/main/java/service/ContestManagementService.java
package service;

import model.*;
import repository.CompetitionRepository;
import repository.OrganiserRepository;
import repository.ParticipantRepository;
import repository.RegistrationRepository;

import java.util.Arrays;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class ContestManagementService {
    private ParticipantRepository participantRepository;

    private CompetitionRepository competitionRepository;

    private OrganiserRepository organiserRepository;

    private RegistrationRepository registrationRepository;

    private Organiser currentOrganiser = null;

    public ContestManagementService(ParticipantRepository participantRepository, CompetitionRepository competitionRepository, OrganiserRepository organiserRepository, RegistrationRepository registrationRepository) {
        this.participantRepository = participantRepository;
        this.competitionRepository = competitionRepository;
        this.organiserRepository = organiserRepository;
        this.registrationRepository = registrationRepository;
    }

    public Organiser getCurrentOrganiser() {
        return currentOrganiser;
    }

    public void login(String username, String password) {
        Organiser o = organiserRepository.find(username);
        if (o == null || !o.getPassword().equals(password))
            throw new ServiceException("Incorrect username or password!");
        currentOrganiser = o;
    }

    public Iterable<String> getCompetitionTypesByString(String str) {
        return competitionRepository.findTypesByString(str);
    }

    public Iterable<String> getAgeCategoriesByCompetitionType(String competitionType) {
        return competitionRepository.findAgeCategories(competitionType);
    }

    public Iterable<Participant> getParticipantsByCompetition(String competitionType, String ageCategory) {
        Competition comp = competitionRepository.findByProps(competitionType, ageCategory);
        return participantRepository.findParticipantsByCompetition(comp);
    }

    public int countParticipantsForCompetition(String competitionType, String ageCategory) {
        Competition comp = competitionRepository.findByProps(competitionType, ageCategory);
        return competitionRepository.countParticipantsForCompetition(comp.getId());
    }

    public Iterable<Long> findParticipantIdsByName(String firstName, String lastName) {
        return participantRepository.findParticipantIdsByName(firstName, lastName);
    }

    public int getParticipantAge(Long participantId) {
        return participantRepository.find(participantId).getAge();
    }

    public Iterable<String> getAgeCategoriesForParticipant(int age, String competitionType) {
        return StreamSupport.stream(getAgeCategoriesByCompetitionType(competitionType).spliterator(), false)
                .filter(s -> {
                    String[] limits = s.split("-");
                    int min = Integer.parseInt(limits[0]);
                    int max = Integer.parseInt(limits[1]);
                    return age >= min && age <= max;
                })
                .collect(Collectors.toList());
    }

    public void addRegistration(Long participantId, String firstName, String lastName, int age,
                                String competitionType, String ageCategory) {
        if (participantId == null)
            participantId = participantRepository.add(new Participant(firstName, lastName, age));

        Participant p = participantRepository.find(participantId);
        Competition c = competitionRepository.findByProps(competitionType, ageCategory);

        String[] ageLimits = c.getAgeCategory().split("-");
        int minAge = Integer.parseInt(ageLimits[0]);
        int maxAge = Integer.parseInt(ageLimits[1]);
        if (age < minAge || age > maxAge)
            throw new ServiceException("Participant age isn't in the allowed range for the competition!");


        Iterable<Registration> registrations = registrationRepository.findByParticipant(participantId);
        int rCount = 0;
        for (var r : registrations) {
            if (r.getId().getLeft().equals(participantId) &&
                    r.getCompetition().getCompetitionType().equals(competitionType) &&
                    r.getCompetition().getAgeCategory().equals(ageCategory))
                throw new ServiceException("Registration already exists!");
            rCount++;
        }
        if (rCount >= 2)
            throw new ServiceException("Participant already registered in 2 competitions!");

        registrationRepository.add(new Registration(p, c));
    }
}

File: contest-management-cs/contest-management-gui-cs/service/ContestManagementService.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Security.Authentication;
using contest_management_gui_cs.model;
using contest_management_gui_cs.repository;

namespace contest_management_gui_cs.service
{
    public class ContestManagementService
    {
        public IParticipantRepository ParticipantRepository { private get; init; }
        public ICompetitionRepository CompetitionRepository { private get; init; }
        public IOrganiserRepository OrganiserRepository { private get; init; }
        public IRegistrationRepository RegistrationRepository { private get; init; }

        public Organiser CurrentOrganiser { get; private set; }

        public void Login(string username, string password)
        {
            var o = OrganiserRepository.Find(username);
            
            if (o == null || o.Password != password)
                throw new InvalidCredentialException("Invalid username or password!");

            CurrentOrganiser = o;
        }

        public IEnumerable<string> GetCompetitionsByString(string competitionType)
        {
            return CompetitionRepository.FindTypesByString(competitionType);
        }

        public IEnumerable<string> GetAgeCategoriesByCompetitionType(string competitionType)
        {
            return CompetitionRepository.FindAgeCategories(competitionType);
        }
        
        public IEnumerable<Participant> GetParticipantsByCompetition(string competitionType, string ageCategory)
        {
            var comp = CompetitionRepository.FindByProps(competitionType, ageCategory);
            return ParticipantRepository.FindParticipantsByCompetition(comp.Id);
        }

        public int CountParticipantsForCompetition(string competitionType, string ageCategory)
        {
            var comp = CompetitionRepository.FindByProps(competitionType, ageCategory);
            var dict = CompetitionRepository.CountParticipantsForEachCompetition();
            foreach (var (key, value) in dict)
            {
                if (key.Id == comp.Id)
                    return value;
            }

            return 0;
        }

        public void AddRegistration(string firstName, string lastName, int age, string competitionType,
            string ageCategory)
        {
            var p = ParticipantRepository.FindParticipantByName(firstName, lastName);
            if (p == null)
            {
                ParticipantRepository.Add(new Participant
                {
                    FirstName = firstName,
                    LastName = lastName,
                    Age = age
                });
                p = ParticipantRepository.FindParticipantByName(firstName, lastName);
            }

            var c = CompetitionRepository.FindByProps(competitionType, ageCategory);

            var r = RegistrationRepository.Find(new Tuple<long, long>(p.Id, c.Id));

            if (r == null)
            {
                var regs = RegistrationRepository.FindByParticipant(p.Id);
                if (regs.Count() < 2)
                    RegistrationRepository.Add(new Registration
                    {
                        RegisteredParticipant = p,
                        WhichCompetition = c
                    });
                else throw new Exception("Participant is already registered to 2 competitions!");
            }
            else throw new DuplicateNameException("Registration already exists!");
        }
    }
}